### ISSUE TITLE
Rule proposal: Prevent moot `unknown` constraint/default in generics

### ISSUE THREAD

**Issue by sindresorhus** [typescript-eslint/typescript-eslint#598]
In TypeScript 3.5, [generic type parameters are implicitly constrained to `unknown` instead of `{}`](https://devblogs.microsoft.com/typescript/announcing-typescript-3-5/#breaking-changes). It would be nice if this rule could warn about moot usage of that.

**Expected Result**

I expected these uses to be reported as unnecessary:

```ts
function set<T extends unknown>(object: T): T {}

function set<T = unknown>(object: T): T {}

set<unknown>(object: {foo: true});
```

**Actual Result**

They are not.

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `1.9.0` |
| `@typescript-eslint/parser`        | `1.9.0` |
| `TypeScript`                       | `3.5.1` |
| `ESLint`                           | `5.16.0` |
| `node`                             | `8.16.0` |
| `npm`                              | `6.9.0` |

**Comment by JoshuaKGoldberg** [comment]
@anthonyronda and I are planning on tackling this issue on Twitch in a few days ✨. https://www.twitch.tv/joshuakgoldberg/schedule?vodID=1414078603

**Comment by JoshuaKGoldberg** [comment]
Actually, adding in an `= unknown` does serve some purpose. Copy & pasting https://github.com/typescript-eslint/typescript-eslint/pull/4658#issuecomment-1079845623:

Ah, actually, our own source code showed why this rule is fine as-is:

```ts
type ObjectLike<T = unknown> = Record<string, T>;

/**
 * Check if the variable contains an object strictly rejecting arrays
 * @param obj an object
 * @returns `true` if obj is an object
 */
function isObjectNotArray<T extends ObjectLike>(
  obj: unknown | unknown[],
): obj is T {
  return typeof obj === 'object' && !Array.isArray(obj);
}
```

Providing an explicit `= unknown` means the generic construct -in this case, `ObjectLike` can be used without an explicit type argument -in this case, `<T>`.

I'll consider this closed as of #4658. Cheers all!
